### DISCUSSION TITLE
How to idiomatically set up code generation?

### DISCUSSION THREAD

**Discussion by pcewing** [premake/premake-core#2051] [Q&A]
I'm trying to understand the idiomatic way in premake to handle build commands
for code generation in a project. As a simple example, consider the
following directory structure:

```
myapp/
|_ build/
|  |_ generated/
|_ src/
|  |_ generator/
|  |  |_ main.cpp
|  |_ myapp/
|  |  |_ main.cpp
|  |_ data/
|     |_ error_codes.json
|_ premake5.lua
```

In this example, the high level build steps for `myapp` would be:

- Build the `generator` application which will be used for code generation
- Run the `generator` application built in the previous step, reading the
  `myapp/src/data/error_codes.json` file as input and writing
  `error_codes.{h,cpp}` files as output to the `build/generated` directory
- Build `build/generated/error_codes.{h,cpp}` as a library project
  called `error_codes`
- Compile the `myapp` executable which includes `build/generated/error_codes.h`
  and links the `error_codes` library project

Omitting all but the important bits for brevity, I imagine the `premake5.lua`
for this would something like:

```
workspace "myapp"
    location "build"

project "generator"
    kind "ConsoleApp"
    files { "src/generator/main.cpp" }

project "error_codes"
    kind "StaticLib"
    includedirs { "build/generated" }
    files { "build/generated/error_codes.cpp" }
    prebuildcommands { "build/generator --input=src/data/error_codes.json --output_dir=build/generated" }

project "myapp"
    kind "ConsoleApp"
    includedirs { "build/generated" }
    files { "src/myapp/main.cpp" }
    links { "error_codes" }
```

There are a few problems with the above:

- Building the `error_codes` project no-ops due to a chicken/egg problem
    - I believe this is because the `build/generated/error_codes.cpp` file
      doesn't exist yet because the pre-build command hasn't run
    - Because the file doesn't exist, the pre-build command doesn't run because
      there's nothing to build
    - I'm not 100% certain this is the reason and still need to confirm
- Moving the pre-build command to be a post-build command of the `generator`
  project sort of works as a hacky workaround but isn't ideal
    - Modifying `error_codes.json` will not result in `error_codes` project
      rebuilding 
    - The only thing that will trigger code to be regenerated is updating the
      generator application

Ideally, I'd want the following criteria to be met:

- If `generator` is rebuilt, rebuild the `error_codes` project
    - Not sure how to declare a dependency like this where a `StaticLib`
      project depends on an `ConsoleApp` project
- If `src/data/error_codes.json` is modified, rebuild the `error_codes` project
    - I thought maybe adding `src/data/*.json` to the `files { ... }` directive
      would accomplish this but it doesn't seem like it 

I feel like the easiest way to accomplish the above would be to leave the
generation step as a pre-build command and make sure it **always** runs but I
don't think there's an option for that.

I feel like code generation is a pretty common build step so I'd be surprised
if I'm the first person to encounter this but I'm struggling to find a
solution.

My only other thought process is to premake the generator separately and do
code generation completely outside premake. Like:

- Premake the `generator` app
- Run the `generator` app
- Premake and build `myapp`

But it would be a lot more convenient to just premake the whole thing at once.

Edit: And after writing this whole thing up it seems like pre-build steps are running consistently for me now. Pretty sure all I did was update to latest version of premake and move the `prebuildcommands` to the end of my `project` definition so maybe this was a bug that was fixed. Closing this out

**Comment by KyrietS** [comment]
I had a number of situations like this when I found a solution to my problem in the process of writing a question on StackOverflow 😄
